opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py

```diff
@@ -511,6 +511,8 @@ def start_span(  # pylint: disable=too-many-locals
         )
 
         if sampling_decision.sampled:
+            options = context.trace_options | trace_api.TraceOptions.SAMPLED
+            context.trace_options = trace_api.TraceOptions(options)
```

Comment:
Technically, we are modifying a context here which is supposed to be immutable. Since we are the unique owner of the context here (we haven't let anyone else see it yet) that's fine, but I wonder if we should make use of `__slots__` and `@property`s in SpanContext.